libraries/standard/mqtt/integration-test/mqtt_system_test.c

```diff
@@ -1027,3 +1060,347 @@ void test_MQTT_Restore_Session_Complete_Incoming_Publish( void )
+    establishMqttSession( &context, &networkContext, false, &persistentSession );
+
+    /* Verify that the session was resumed. */
+    TEST_ASSERT_TRUE( persistentSession );
+
+    /* Obtain the packet ID of the PUBLISH packet that didn't complete in the previous connection. */
+    MQTTStateCursor_t cursor = 0u;
+    uint16_t publishPackedId = MQTT_PublishToResend( &context, &cursor );
+    TEST_ASSERT_EQUAL( context.outgoingPublishRecords[ 0 ].packetId, publishPackedId );
+
+    /* Resend the PUBLISH packet that didn't complete in the previous connection. */
+    TEST_ASSERT_EQUAL( MQTTSuccess, publishToTopic(
+                           &context, TEST_MQTT_TOPIC, true, MQTTQoS2, publishPackedId ) );
+
+    /* Complete the QoS 0 PUBLISH resend operation. */
+    TEST_ASSERT_EQUAL( MQTTSuccess,
+                       MQTT_ProcessLoop( &context, 2 * MQTT_PROCESS_LOOP_TIMEOUT_MS ) );
+
+    /* Make sure that the QoS 2 PUBLISH re-transmission was complete. */
+    TEST_ASSERT_TRUE( receivedPubRec );
+    TEST_ASSERT_TRUE( receivedPubComp );
+
+    /* Make sure that the library has removed the record for the outgoing PUBLISH packet. */
+    TEST_ASSERT_EQUAL( MQTT_PACKET_ID_INVALID, context.outgoingPublishRecords[ 0 ].packetId );
+}
+
+/**
+ * @brief Verifies the behavior of the MQTT library on receiving a duplicate
+ * QoS 1 PUBLISH packet from the broker in a restored session connection.
+ * Tests that the library responds with a PUBACK to for the duplicate incoming QoS 1 PUBLISH
+ * packet that was un-acknowledged in a previous connection of the same session.
+ */
+void test_MQTT_MQTT_Restore_Session_Duplicate_Incoming_Publish_Qos1( void )
+{
+    /* Terminate TLS session and TCP connection network connection to discard current MQTT session
+     * that was created as a "clean session". */
+    ( void ) Openssl_Disconnect( &networkContext );
+
+    /* Establish a new MQTT connection over TLS with the broker with the "clean session" flag set to 0
+     * to start a persistent session with the broker. */
+
+    /* Create the TLS+TCP connection with the broker. */
+    TEST_ASSERT_EQUAL( OPENSSL_SUCCESS, Openssl_Connect( &networkContext,
+                                                         &serverInfo,
+                                                         &opensslCredentials,
+                                                         TRANSPORT_SEND_RECV_TIMEOUT_MS,
+                                                         TRANSPORT_SEND_RECV_TIMEOUT_MS ) );
+    TEST_ASSERT_NOT_EQUAL( -1, networkContext.socketDescriptor );
+    TEST_ASSERT_NOT_NULL( networkContext.pSsl );
+
+    /* Establish a new MQTT connection for starting a persistent session with the broker
+     * by setting the "clean session" flag to 0. */
+    establishMqttSession( &context, &networkContext, false, &persistentSession );
+    TEST_ASSERT_FALSE( persistentSession );
```

Comment:
I'm seeing this pattern a lot of Disconnect TCP -> Connect TCP -> Establish persistent MQTT session, and another below of Reestablish TCP -> Establish persistent MQTT session. Why not add functions like `initializePersistantSession()` and `resumePersistentSession()` so this code doesn't have to be repeated so many times?

Reply:
Good suggestion of commonization and names